### ISSUE TITLE
flatten switching the inner stream to the same inner, not pure

### ISSUE THREAD

**Issue by staltz** [staltz/xstream#90]
https://github.com/cyclejs/cyclejs/issues/365

**Comment by Hypnosphi** [comment]
I'm not really sure it's only `flatten`'s issue. A simple reproducing example is

```js
const a$ = xs.merge(xs.of(1), xs.periodic(1000));
const b$ = a$.fold((acc, x) => acc + x, 0); // 01--1--2--4--
const c$ = a$.fold((acc, x) => acc + x, 0); // 0--0--1--3--
```

In fact, I don't think that real purity is achievable in js

**Comment by staltz** [comment]
That's a separate issue. We could make another one for that.

**Comment by staltz** [comment]
Well actually, if you `merge` b$ and c$, you might not get a bug. Maybe the issue is just that `a$` needed a `remember()` suffixed.

**Comment by Hypnosphi** [comment]
I don't see what would `merge` "fix" here: both b$ and c$ will become listeners of a$, and only one of them will get the synchronous value

**Comment by staltz** [comment]
Right, correct.

**Comment by staltz** [comment]
Fixed in v5.3.2

**Comment by Hypnosphi** [comment]
By the way, two calls of `fold` will produce two different stream (because `fold` is really not pure, as virtually all the other operators aren't). So the sameness check will fail here.
